sweep/store.go

```
@@ -0,0 +1,234 @@
+	utxnFinalizedKndrTxnKey = []byte("finalized-kndr-txn")
+)
+
+// SweeperStore stores published txes.
+type SweeperStore interface {
+	// IsOurTx determines whether a tx is published by us, based on its
+	// hash.
+	IsOurTx(hash chainhash.Hash) (bool, error)
+
+	// NotifyPublishTx signals that we are about to publish a tx.
+	NotifyPublishTx(*wire.MsgTx) error
+
+	// GetLastPublishedTx returns the last tx that we called NotifyPublishTx
+	// for.
+	GetLastPublishedTx() (*wire.MsgTx, error)
+}
+
+type sweeperStore struct {
+	db *channeldb.DB
+}
+
+// NewSweeperStore returns a new store instance.
+func NewSweeperStore(db *channeldb.DB) (SweeperStore, error) {
+	err := db.Update(func(tx *bbolt.Tx) error {
+		_, err := tx.CreateBucketIfNotExists(
+			lastTxBucketKey,
+		)
+		if err != nil {
+			return err
+		}
+
+		if tx.Bucket(txHashesBucketKey) != nil {
+			return nil
+		}
+
+		txHashesBucket, err := tx.CreateBucket(txHashesBucketKey)
+		if err != nil {
+			return err
+		}
+
+		// Use non-existence of tx hashes bucket as a signal to migrate
+		// nursery finalized txes.
+		migrateTxHashes(tx, txHashesBucket)
+
+		return err
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return &sweeperStore{
+		db: db,
+	}, nil
+}
+
+// migrateTxHashes migrates nursery finalized txes to the tx hashes bucket. This
+// is not implemented as a database migration, to keep the downgrade path open.
+func migrateTxHashes(tx *bbolt.Tx, txHashesBucket *bbolt.Bucket) error {
+
+	// Iterate over all nursery chain buckets.
+	return tx.ForEach(func(name []byte, chainBucket *bbolt.Bucket) error {
+		// Skip non-nursery buckets
+		if len(name) < len(utxnChainPrefix) {
+			return nil
+		}
+
+		if !bytes.Equal(name[:len(utxnChainPrefix)], utxnChainPrefix) {
+			return nil
+		}
+
+		// Retrieve the existing height index.
+		hghtIndex := chainBucket.Bucket(utxnHeightIndexKey)
+		if hghtIndex == nil {
+			return nil
+		}
+
+		// Retrieve all heights.
+		hghtIndex.ForEach(func(k, v []byte) error {
```

Ignoring the error here and a few places below in this method. However the next diff (IMO) should move this into the nursery so we don't need to spill all the information of its bucket structure over.

---

Error handled. Just saw it one more times, fixed that too. Info spill, see comments above.